### DISCUSSION TITLE
How to get all UnhandledExceptions?

### DISCUSSION THREAD

**Discussion by AuderixDev** [unoplatform/uno#19070] [Q&A]
Hi all, 

what are the best practices to catch/handle all exceptions?

I want to notify the user about all exceptions. 
It doesn't matter if they are thrown by me, are handled or unhandled. 
I want to catch all exceptions in the main thread, in tasks, in manged or unmanged code.

```
App.Current.UnhandledException += (s, e) => { HandleException(e.Exception); e.Handled = true; };
AppDomain.CurrentDomain.UnhandledException += (s, e) => { HandleException(e.ExceptionObject as Exception); };
TaskScheduler.UnobservedTaskException += (s, e) => { HandleException(e.Exception.InnerException); e.SetObserved(); };
```

The problem with this code: 
In WASM not a single exception is caught. 
In WinAppSDK exceptions from the main thread are caught, but exceptions from Tasks (TaskScheduler) appears only 60 (?)seconds later.
(iOS and Android not tested)

There is another way:
`AppDomain.CurrentDomain.FirstChanceException += (s, e) => { HandleException(e.Exception); };`

This works really well in WASM and WinAppSDK to catch all exceptions. 
But it catches also a lot of Warnings from WASM that keeps appearing in the Browser Console:
![image](https://github.com/user-attachments/assets/8cac1009-be7b-4336-a36f-05e5590e7175)

**Comment by jeromelaban** [comment]
Thanks for the report. 

It looks like you're trying to use Wasm Threading, which is experimental at this time in net8 and not working in net9. See [this documentation](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/using-the-bootstrapper.html#threading) for more details.

**Reply by AuderixDev** [reply]
HI @jeromelaban 

Thank you for the information about wasm threading. 
But this didn't answer the question - what is the best way to catch exceptions?

`AppDomain.CurrentDomain.FirstChanceException` has also the disadvantage, that it is triggered for every rethrow of a exception. 
But all other Exceptionsevents are not triggered for an Exception in an async Command.

**Reply by jeromelaban** [reply]
For exceptions, this may be specific to the scenario that is throwing the exception. You're mentioning async commands, are you using MVUX or something else?

**Reply by AuderixDev** [reply]
Yes, I'm using MVUX.
But I don't want to try/catch every Command or create a "Base"-Command that catches the exception. 
In WPF I can subscribe to Application.Current.Dispatcher.UnhandledException and I get all Exceptions. 

I want to handle the exceptions globally so that I only have the implementation in one place. (logic for logging, inform user)

**Reply by dr1rrb** [reply]
@AuderixDev  You can set [`Uno.Extensions.Reactive.Command.DefaultErrorHandler`](https://github.com/unoplatform/uno.extensions/blob/main/src/Uno.Extensions.Reactive/Presentation/Commands/Command.cs#L23C34-L23C53) for this.

**Reply by AuderixDev** [reply]
@dr1rrb Thanks that is definitely helpful! (and something that is not documented?)
So, MVUX Command Exceptions are handled most of the time. 
but as soon as I execute something asynchronous (like an API call), but do not await it, the exception is not caught.

Tested with this code:
```
public async Task ThrowTestException()
{
        Task.Run(() => throw new Exception("Test exception"));
}
```

(if I await the Task.Run, the exception is caught)

**Reply by FrozDark** [reply]
It's expected as that Task is executed outside of the context and not observed with await
